### PR TITLE
Adaptive Batching v2

### PR DESCRIPTION
**Pull request by Corey-Zumar**
TODO: Add tests

### CODE REVIEW

**Comment by dcrankshaw** [comment]
@Corey-Zumar Is this ready for review yet?

**Comment by Corey-Zumar** [comment]
@dcrankshaw I'll add some documentation and have it ready for review within the hour. Tests still need to be written. I've merged the branch associated with #393 into this PR; that explains the large diff. This will be easier to review once #393 is merged.

File: src/libclipper/src/containers.cpp [hunk 1]
```diff
@@ -30,78 +34,191 @@ ModelContainer::ModelContainer(VersionedModelId model, int container_id,
...
+  if (mb_search != processing_datapoints_.end() &&
+      std::get<0>(mb_search->second) < MINIMUM_BATCH_SAMPLE_SIZE) {
+    // We don't have a large enough latency sample
+    // corresponding to the maximum batch size, so
+    // we won't update the maximum size
+    return max_batch_size_;
+  } else if (max_batch_size_ < ADDITIVE_EXPANSION_THRESHOLD) {
+    return max_batch_size_ + 1;
+  } else {
+    double expansion_factor = exploration_distribution_(exploration_engine_);
```
Comment (dcrankshaw) [inline src/libclipper/src/containers.cpp:227]: Make sure that `expansion_factor` is non-negative (you can do a `max(0, expansion_factor)`).

File: src/libclipper/src/containers.cpp [hunk 2]
```diff
@@ -30,78 +34,191 @@ ModelContainer::ModelContainer(VersionedModelId model, int container_id,
       latency_hist_("container:" + model.serialize() + ":" +
                         std::to_string(replica_id) + ":prediction_latency",
                     "microseconds", HISTOGRAM_SAMPLE_SIZE),
-      avg_throughput_per_milli_(0),
-      throughput_buffer_(THROUGHPUT_BUFFER_CAPACITY) {
+      max_batch_size_(1),
+      max_latency_(0),
+      explore_dist_mu_(.1),
+      explore_dist_std_(.05),
+      budget_decay_(.92),
```
Comment (dcrankshaw) [inline src/libclipper/src/containers.cpp]: Add comments about how these numbers were chosen (even if that is just that you did some offline analysis and picked numbers that worked reasonably well).

File: src/libclipper/include/clipper/task_executor.hpp [hunk 3]
```diff
@@ -138,18 +138,29 @@ class ModelQueue {
...
+    if (batch_differential > 0) {
+      // Artificially inject queries to create
+      // a full batch
+      //
+      // Create a copy of the last task in the batch and make it artifical
+      PredictTask last_task = batch.back();
+      last_task.artificial_ = true;
+
+      std::fill_n(std::back_inserter(batch), batch_differential, last_task);
+    }
```
Comment (dcrankshaw) [inline src/libclipper/include/clipper/task_executor.hpp]: We should only inject artificial padding queries if we are exploring. The idea would be that once the system has explored sufficiently and is running at steady state, no more artificial queries are being injected.

File: src/libclipper/src/containers.cpp [hunk 4]
```diff
@@ -30,78 +34,191 @@ ModelContainer::ModelContainer(VersionedModelId model, int container_id,
...
-    batch_size = 1;
+  size_t curr_batch_size;
+  if (budget > static_cast<double>(max_latency_)) {
+    curr_batch_size = explore();
+  } else {
+    curr_batch_size = estimate(budget);
   }
-  return batch_size;
+  max_batch_size_ = std::max(curr_batch_size, max_batch_size_);
+  return curr_batch_size;
```
Comment (dcrankshaw) [inline src/libclipper/src/containers.cpp]: Return an additional flag that indicates whether this batch size was selected using exploration or estimation.

**Review by dcrankshaw — APPROVED** [review]
Make the changes we discussed, then merge this.

Some notes:
+ Link to the adaptive batching slides (or better yet, write them up in a markdown file) and put it somewhere that makes sense (probably in src/libclipper/...)

File: src/libclipper/src/containers.cpp (hunk 1 above)
Comment (Corey-Zumar) [reply src/libclipper/src/containers.cpp:227]: done

**Comment by Corey-Zumar** [comment]
@dcrankshaw If you have time to take a look, dlib needs to be included as a static library rather than an interface library (the debug build fails with the interface library setup). I was running into export issues at link time when attempting to link a statically-defined dlib to libclipper.

**Comment by dcrankshaw** [comment]
Okay I'll take a look

**Comment by dcrankshaw** [comment]
@Corey-Zumar fixed the static lib issue. I'll keep on eye on the tests tonight and try to get this merged. Are there any known outstanding issues or are we just waiting for the PR to get through jenkins?

**Comment by dcrankshaw** [comment]
@Corey-Zumar Do you still need to write tests for this? Jenkins is failing because `container_tests.cpp` is calling functions that no longer exist.

**Comment by Corey-Zumar** [comment]
@dcrankshaw yes, I still need to write a couple tests. Will aim to do that tonight.

**Comment by Corey-Zumar** [comment]
jenkins test this please

**Comment by dcrankshaw** [comment]
jenkins test this please

**Comment by dcrankshaw** [comment]
jenkins test this please

**Pull request merged**